server/core/session.cc

```diff
@@ -278,6 +279,9 @@ void session_link_backend_dcb(MXS_SESSION *session, DCB *dcb)
     dcb->service = session->service;
     /** Move this DCB under the same thread */
     dcb->poll.thread.id = session->client_dcb->poll.thread.id;
+    /** Track session's backends */
+    dcb->next_backend = session->backends.head;
```

Comment:
This list must be updated every time a DCB is closed as otherwise the pointer will point to freed memory.
The correct place to do it seems to be in `dcb_final_close` inside the following branch.

```c++
    if (dcb->dcb_role == DCB_ROLE_BACKEND_HANDLER &&  // Backend DCB
        dcb->state == DCB_STATE_POLLING &&            // Being polled
        dcb->persistentstart == 0 &&                  /** Not already in (> 0) or being evicted from (-1)
                                                       * the persistent pool. */
        dcb->server)                                  // And has a server
    {
        /* May be a candidate for persistence, so save user name */
        const char *user;
        user = session_get_user(dcb->session);
        if (user && strlen(user) && !dcb->user)
        {
            dcb->user = MXS_STRDUP_A(user);
        }

       ///////////////////
       // ADD CODE HERE //
       ///////////////////

        if (dcb_maybe_add_persistent(dcb))
        {
            dcb->n_close = 0;
        }
    }
```

Creating a new function (e.g. `void session_unlink_backend_dcb(MXS_SESSION *session, DCB *dcb)`) that updates the list would solve the problem.